Tootbot/Source/Controllers/ImageCacheController.swift

```diff
@@ -0,0 +1,116 @@
+//
+// Copyright (C) 2017 Tootbot Contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Foundation
+import Alamofire
+import ReactiveSwift
+
+class ImageCacheController {
+
+    enum Error: Swift.Error {
+        case downloadFailed(Swift.Error?)
+        case writeToFileFailed(Swift.Error?)
+    }
+
+    private static let memoryCache = NSCache<NSString, UIImage>()
+    private static let cacheDirectoryURL: URL = {
+        let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
+        let url = URL(fileURLWithPath: documentsPath)
+        return url.appendingPathComponent("ImageCache")
+    }()
+
+    private let fileManager = FileManager.default
+
+    init() {
+        if fileManager.fileExists(atPath: ImageCacheController.cacheDirectoryURL.path) == false {
+            try? fileManager.createDirectory(at: ImageCacheController.cacheDirectoryURL,
+                                        withIntermediateDirectories: true)
+        }
+    }
+
+    public func fetch(url: URL) -> SignalProducer<UIImage, Error> {
+        let key = self.key(for: url)
+        return downloadImage(from: url)
```

Comment:
Add a `.take(first: 1)` to the end, which makes the resulting signal producer complete after the first value.